### ISSUE TITLE
Librosa in 128MB memory or less

### ISSUE THREAD

**Issue by Cortexelus** [librosa/librosa#406]
I want to package librosa onto AWS Lambda (a serverless compute service) to do audio analysis (similar to EchoNest API) cheaply and scalably. To minimize cost and maximize performance, I want to use as little memory as possible. 128MB is ideal. But I also want to support the analysis of DJ mixes that are an hour+ long. 

If I partially transcode->STFT->etc n samples at a time, I can still do mostly what I need: HPSS, cqt, transient-segmentation, timbre, pitch, or time stretch.

Let's say the input is a URL to an audio file (could be MP3, M4A/AAC, WAV, FLAC)

1. Partially download n bytes
2. Send input through the transcode->STFT->etc pipeline 
2. `del` bytes for garage collection when no longer needed
3. Print output as I go (instead of gathering into a dictionary)
4. Hop to next bytes 

Byte range needs to take into consideration the HPSS values (and segmentation boundaries) at frame f depend on a few frames before and after it.

Aware of anything stubborn/tricky about transcoding partially downloaded files?

Recommendation on how I might best structure/integrate this into librosa? Perhaps something in examples/ 

Recommendations for memory profilers?

**Comment by bmcfee** [comment]
> Aware of anything stubborn/tricky about transcoding partially downloaded files?

Dunno - I've never tried it.  You might want to try pysoundfile for this, but I'm not sure how well it supports streaming.  Failing that, I think this is exactly the kind of thing gstreamer is designed for, but to do it right, you might need to work directly with pygobject.  The audioread backend for gstreamer might be a good source to pull from here.

Alternately, you could try hooking your stream through portaudio and using pyaudio to grab frames for analysis.  This way, you separate the decoding logic from the streaming and analysis.  I've done this kind of thing before to have librosa analyze audio from the mic through jupyter notebook ([ipyaudio](https://github.com/bmcfee/ipyaudio), which no longer works due to ipython/jupyter api chainges).  The latency there isn't pretty, but it works.

**Comment by dela3499** [comment]
My first thought would be to download the entire file, split it into chunks with [pydub](https://github.com/jiaaro/pydub), and then run whatever analysis you like on the chunks (perhaps [Dask](http://dask.pydata.org/en/latest/) would be helpful here.

**Comment by Cortexelus** [comment]
I could download the entire file, but then I must set a maximum upload
filesize. A 10 minute WAV file is already 100MB.

If it streams, I don't wait for the entire download, I begin analyzing
immediately, which is faster.

**Comment by bmcfee** [comment]
Coming back to this one, after quite some time:

I don't think there's an easy solution to using librosa in severely limited resource environments.  We're making a bit of progress in #157 which cuts down on some intermediate variables / memory copies generated by vectorization, but I think the general problem is a bit above our pay grade here.

A big part of the problem here is that early on, we made a hard decision to focus on batch processing rather than streaming.  This was motivated by our specific needs at the time, and a preference to keep the API and underlying components as simple as possible.  (Meaning: i didn't want to force users to learn how to wire together a bunch of stateful objects just to get a mel spectrogram.)

Now, you can hack it by iterating over short buffers and processing chunks as you go, but this will only work for certain functions.  (For instance, CQT wouldn't work here.)

One potential, long-term solution, would be to revamp the guts of librosa to use stateful objects which can iterate over frames / buffers, but provide a thin layer of wrappers (ie the current API) that hides this from the user (unless they want to use the object api directly).  I would consider this a 2.0 milestone feature, as it's pretty ambitious.

This is all to say: I don't think we can do much here in the short term, and I'd like to close this issue out.  Any objections?

**Comment by misrasaurabh1** [comment]
How did you even make librosa even work on AWS lambda? How did you package the linux libav-tools dependency on Lambda?

**Comment by deepwilson** [comment]
Yeah, please do let us know how you managed to squeeze librosa into an AWS package !

**Comment by tblake84** [comment]
I have not run into any issues packaging librosa into a lambda package and it stays under the 262MB hard limit on extracted function size (I honestly don't know how, it just does).  However, librosa.load is giving me audioread.NoBackendError exceptions so I am dealing with that now.

**Comment by dhemgren** [comment]
@tblake84 have you made any progress? we've not managed to make it stay under the 250 mb limit. for us, the dependencies alone (numpy, sklearn, scipy in particular) approach this limit so would be great if you could point me in the right direction! appreciate any help :)

**Comment by Jasparr77** [comment]
I've found a promising approach using [Zappa](https://github.com/Miserlou/Zappa) and Flask, Zappa has a 'slim_handler' which helps to cut down on size. 
That being said, I'm currently running into issues with soundfile not installing correctly. Promising, though! Will reply back if I solve it.

**Comment by yousefjnm** [comment]
@Jasparr77 any more good news?

**Comment by Jasparr77** [comment]
@yousefjnm Nothing yet, I'm working through a few issues specific to my use-case. I would recommend trying Zappa out, though - it's a pretty straightforward solution and the [slim_handler](https://github.com/Miserlou/zappa-blog/blob/master/posts/slim-handler.md) feature leverages s3 to "support up to 500M of zipped up Python projects."
